README
NAME
    Data::Handle - A Very simple interface to the __DATA__ file handle.

VERSION
    version 1.000001

SYNOPSIS
        package Foo;

        sub bar {
            my $handle = Data::Handle->new( __PACKAGE__ );
            while (<$handle>) {
                print $_;
            }
        }

        __DATA__
        Foo

DESCRIPTION
    This Package serves as a very *very* simple interface to a packages
    __DATA__ section.

    Its primary purposes is to make successive accesses viable without
    needing to scan the file manually for the __DATA__ marker.

    It does this mostly by recording the current position of the file handle
    on the first call to "->new", and then re-using that position on every
    successive "->new" call, which eliminates a bit of the logic for you.

    At present, it only does a simple heuristic ( backtracking ) to verify
    the current position is immediately at the start of a __DATA__ section,
    but we may improve on this one day.

METHODS
  new
        my $fh = Data::Handle->new( $targetpackage )

    Where $targetpackage is the package you want the __DATA__ section from.

WARNING
    At present, this module does you no favors if something else earlier has
    moved the file handle position past the __DATA__ section, or rewound it
    to the start of the file. This is an understood caveat, but nothing else
    seems to have a good way around this either. ( You can always rewind to
    the start of the file and use heuristics, but that is rather pesky ).

    Hopefully, if other people do decide to go moving your file pointer,
    they'll use this module to do it so you your code doesn't break.

USAGE
    "Data::Handle-"new()> returns a tied file-handle, and for all intents
    and purposes, it should behave as if somebody had copied __DATA__ to its
    own file, and then done "open $fh, '<' , $file" on it, for every
    instance of the Data::Handle.

    It also inherits from IO::File, so all the methods it has that make
    sense to use should probably work on this too, i.e.:

        my $handle = Data::Handle->new( __PACKAGE__ );
        my @lines = $handle->getlines();

    Also, all offsets are proxied in transit, so you can treat the
    file-handle as if byte 0 is the first byte of the data section.

        my $handle = Data::Handle->new( __PACKAGE__ );
        my @lines = $handle->getlines();
        seek $handle, 0, 0;
        local $/ = undef;
        my $line = scalar <$handle>; # SLURPED!

    Also, the current position of each handle instance is internally
    tracked, so you can have as many objects pointing to the same __DATA__
    section but have their read mechanism uninterrupted by any others.

        my $handlea  = Data::Handle->new( __PACKAGE__ );
        my $handleb  = Data::Handle->new( __PACKAGE__ );

        seek $handlea, 10, 0;
        seek $handleb, 15, 0;

        read $handlea, my $buf, 5;

        read $handleb, my $bufa, 1;
        read $handleb, my $bufb, 1;

         $bufa eq $bufb;

    Don't be fooled, it does this under the covers by a lot of "seek"/"tell"
    magic, but they shouldn't be a problem unless you are truly anal over
    speed.

CREDITS
    Thanks to LeoNerd and anno, from #perl on irc.freenode.org, they were
    most helpful in helping me grok the magic of "tie" that makes the
    simplicity of the interface possible.

    Thanks to Chas Owens and James Wright for their efforts with trying to
    get something simpler with fdup()ing the descriptor ( Sadly not working
    yet ).

AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric <kentnl@cpan.org>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Data-Handle

1.000002 2017-03-08T11:45:07Z ded6738

1.000001 2017-03-08T11:07:34Z 18a81a7
 - Bugfix: Avoid test failures on -Ddefault_inc_excludes_dot
 - Lower Test::More requirements

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +11 ↑3 -3 (recommends: +1, suggests: ↑2)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.000000 2014-10-10T13:04:03Z 3f54592
 [00 Major - Version Scheme Change]
 - Migrate to using x.yyyzzz (mantissa=6) instead of x.yyyzzzaa (mantissa=8)
 - This may affect hand-operated downstreams
 - https://questhub.io/realm/perl/quest/5388d3a9bbd0be2224000129

 [01 Major - Toolkit Change]
 - Migrate from using Module::Build to using ExtUtils::MakeMaker as installer toolkit
 - Make test/toolkit "latest" dependencies merely recommendations.
 - https://questhub.io/realm/perl/quest/5388cc12bbd0be5b2500011f

 [Dependencies::Stats]
 - Dependencies changed since 0.02001003, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +47 -2 (recommends: -1, suggests: +1 ↑1)
 - test: +2 ↓1 -2 (recommends: +3)

0.02001003 2013-07-26T09:11:29Z
 [Bugfixes]
 - No longer broken on modern Package::Stash's

 [Dependencies::Stats]
 - Dependencies changed since 0.02001002, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +6 (suggests: ↑1)
 - runtime: ↑1
 - test: +1

 [Documentation]
 - Updated copyright year
 - Added README.mkdn

 [Meta]
 - Bug tracker moved to github

 [Packaging]
 - use test_requires (mb 0.4004)

0.02001002 2012-08-11T01:59:03Z
 - Minor Maintenance Release.

 [Dependencies::Stats]
 - Dependencies changed since 0.02001001, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: (suggests: ↑1)
 - test: ↑1

 [Tests]
 - Made tests slightly more detailed in the error condition.

0.02001001 2012-06-08T13:30:28Z
 [00 Important Changes]
 - Code has been purged of stolen chunks of Carp code, and now instead depends on a minimum Carp of v1.22. This is the
   first release of Carp that was packaged on CPAN as a Dual Lifed module, upon which I can rely upon the existance and
   implementation of 'caller_info'. It is hoped this will make bugs smaller and the code easier.

 [Dependencies::Stats]
 - Dependencies changed since 0.01011704, see misc/*.deps* for details
 - develop: (recommends: ↑1, suggests: ↑1)
 - runtime: +1 ↑1

 [Packaging]
 - Update LICENSE ( Year, Address, Indent )

 [Tests::New]
 - minimum version test

0.01011704 2011-10-15T11:05:19Z
 [Bugs]
 - Adapted to work with latest Package::Stash ( 0.33 )

 [Dependencies::Stats]
 - Dependencies changed since 0.01011703, see misc/*.deps* for details
 - runtime: +3
 - test: -1

 [Internals]
 - AUTHORITY data now provided
 - Versions not set in BEGIN
 - Tidy step now only 80 chars.

 [Packaging]
 - Updated LICENSE ( Better indentation, updated address for FSF )
 - Author/Release tests now say in xt/
 - Github URL's now all https://

 [Tests]
 - Refactored some of the exception class testing logic in 04_ for enhanced clarity.

0.01011703 2011-04-24T22:29:18Z
 - Minor changes release, mostly just dist maintainence.
 - [Dependencies::Stats]
 - Dependencies changed since 0.01011702, see misc/*.deps* for details
 - develop: +1 -1 (recommends: +1 -1, suggests: +1 -1)

 [Dependencies::Stats]
 - Dependencies changed since 0.01011702, see misc/*.deps* for details
 - develop: +1 -1 (recommends: +1 -1, suggests: +1 -1)

 [Internals]
 - Explicitly set $@ to undef ( critic )
 - Stop setting $DB::VERSION

 [Packaging]
 - Move from @KENTNL to @Author::KENTNL
 - Use recent features of @Author::KENTNL
 - Ship .perltidyrc
 - Changes file normalized to CPAN::Changes form.

 [Tests]
 - Add CPAN::Changes tests.
 - Add release portability tests.

0.01011702 2010-12-21T13:43:35Z
 - Distribution Change only. Otherwise identical to previous release.
 - Don't index DB.

0.01011701 2010-12-21T13:29:00Z
 - Give up on trying to operate with the Carp shipped with 5.6.2. If its not there, and not in Carp::Heavy's
   injections, use a copy of one we stole and made work.
 - Have to put a fix in as well, because Term::ANSIColor clobbers $@

0.01011617 2010-12-21T05:46:24Z
 - Fix "missing parentheses around my ( )" for Perl-5.6

0.01011601 2010-12-20T13:07:01Z
 - Yuck. The way caller_info is provided is ambiguous. Hopefully this release will get that right.

0.01011501 2010-12-19T13:15:59Z
 - The method we were borrowing out of Carp wasn't there in old versions, so we've added a bit of magic to try get it
   from older carps too, so we don't fail on <= 5.10.1 :(

0.01011500 2010-12-19T12:36:06Z
 - More tests to improve general function coverage.
 - 'generate_exception' is gone, depercated.
 - Replace: C< generate_exception($class , $message)->throw() > with C< _gen_tree($full_class)->throw( $message ) >
 - Coverage testing detected a missing exception class, fixed in this release.
 - All tied() methods now tested to some degree to work.

0.01011421 2010-12-19T09:52:17Z
 - Reworked the Exception code to be cleaner at the cost of a slight initialisatin cost.
 - Reworked code bodies to unify seek/tell calls as _restore_pos and _set_pos.
 - Added _fh so internal code is less confusing.
 - Extended tests to cover more scenarios.
 - Added contributed suggestions as "alternative_techinques" with demonstrations how they *( so far ) don't work. If
   people can get them to work (on all perls), that'd be great!. ( Seems there are many weird fdup bugs in perl )
 - Stolen the backtrace code from Carp a bit and augmented it, now produce useful backtraces.
 - Have some nice code in our exception stringification that highlights lines with colours based on how much we think
   they're related to the problem. Stuff thats part of DH ( the place the error was detected ) is highlighted green as
   we think it will be usefull in diagnosis, but its not the source of the problem. Stuff that we know is likely to be

Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "A Very simple interface to the __DATA__  file handle.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Data-Handle",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Data::Handle",
  "PREREQ_PM" => {
    "Carp" => "1.22",
    "IO::File" => 0,
    "Package::Stash" => "0.15",
    "Scalar::Util" => 0,
    "Symbol" => 0,
    "Term::ANSIColor" => 0,
    "Try::Tiny" => 0,
    "overload" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "1.000002",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/alternative_techniques/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => "1.22",
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::File" => 0,
  "IO::Handle" => 0,
  "Package::Stash" => "0.15",
  "Scalar::Util" => 0,
  "Symbol" => 0,
  "Term::ANSIColor" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "Try::Tiny" => 0,
  "lib" => 0,
  "overload" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Data/Handle.pm
use 5.006;
use strict;
use warnings;

package Data::Handle;

our $VERSION = '1.000002';

# ABSTRACT: A Very simple interface to the __DATA__  file handle.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY





















































































my %datastash;
use Symbol qw( gensym );
use Scalar::Util qw( weaken );
use parent qw( IO::File );
use Package::Stash 0.15;    # has_symbol
use Carp ();
use Data::Handle::Exception;
use Data::Handle::IO;
use Try::Tiny qw( try catch );









sub new {
  my ( $class, $targetpackage ) = @_;

  _e('NoSymbol')->throw("$targetpackage has no DATA symbol")
    if ( !$class->_has_data_symbol($targetpackage) );

  if ( !$class->_is_valid_data_tell($targetpackage) ) {
    _e('BadFilePos')
      ->throw( "$targetpackage has a DATA symbol, but the filepointer"
        . " is well beyond the __DATA__ section.\n"
        . " We can't work out safely where it is.\n"
        . $class->_stringify_metadata($targetpackage)
        . "\n" );
  }

  my $sym  = gensym();
  my $xsym = $sym;
  weaken($xsym);

  ## no critic( ProhibitTies )
  tie *{$sym}, 'Data::Handle::IO', { self => $xsym };
  ${ *{$sym} }{stash} = {};
  bless $sym, $class;
  $sym->_stash->{start_offset}   = $class->_get_start_offset($targetpackage);
  $sym->_stash->{targetpackage}  = $targetpackage;
  $sym->_stash->{current_offset} = $class->_get_start_offset($targetpackage);
  $sym->_stash->{filehandle}     = $class->_get_data_symbol($targetpackage);
  return $sym;

}

sub _has_data_symbol {
  my ( undef, $package ) = @_;
  my $rval = undef;
  try {
    my $stash = Package::Stash->new($package);
    return unless $stash->has_symbol('DATA');
    my $fh = $stash->get_symbol('DATA');
    $rval = defined fileno *{$fh};
  }
  catch {
    if (/is not a module name/) {
      $rval = undef;
      return;
    }
    ## no critic (RequireCarping)
    die $_;
  };
  return $rval;
}

sub _get_data_symbol {
  my ( $self, $package ) = @_;
  if ( !$self->_has_data_symbol($package) ) {
    _e('Internal::BadGet')->throw('_get_data_symbol was called when there is no data_symbol to get');
  }
  return Package::Stash->new($package)->get_symbol('DATA');
}

sub _get_start_offset {
  my ( $self, $package ) = @_;

  return $datastash{$package}->{offset}
    if ( exists $datastash{$package}->{offset} );

  if ( !$self->_has_data_symbol($package) ) {
    _e('Internal::BadGet')->throw('_get_start_offset was called when there is no data_symbol to get');
  }
  my $fd       = $self->_get_data_symbol($package);
  my $position = tell $fd;

  $datastash{$package}->{offset} = $position;

  return $position;
}

sub _is_valid_data_tell {
  my ( $self, $package ) = @_;
  return 1
    if ( exists $datastash{$package} && 1 == $datastash{$package}->{valid} );
  if ( !$self->_has_data_symbol($package) ) {
    _e('Internal::BadGet')->throw('_is_valid_data_tell was called when there is no data_symbol to get');
  }

  my $fh     = $self->_get_data_symbol($package);
  my $offset = $self->_get_start_offset($package);

  # The offset to the start of __DATA__ is 9 bytes because it includes the
  # trailing \n
  #
  my $checkfor = qq{__DATA__\n};
  seek $fh, ( $offset - length $checkfor ), 0;
  read $fh, my ($buffer), length $checkfor;
  seek $fh, $offset, 0;

  $datastash{$package}->{previous_bytes} = $buffer;

  if ( $buffer eq $checkfor ) {
    $datastash{$package}->{valid} = 1;
    return 1;
  }
  else {
    $datastash{$package}->{valid} = 0;
    return;
  }
}

sub _stringify_metadata {
  my ( undef, $package ) = @_;
  my @lines = ();
  if ( not exists $datastash{$package} ) {
    push @lines, "Nothing known about $package\n";
    return join "\n", @lines;
  }
  else {
    push @lines, q{Offset : } . $datastash{$package}->{offset};
    push @lines, q{Prelude : '} . $datastash{$package}->{previous_bytes} . q{'};
    push @lines, q{Valid: } . $datastash{$package}->{valid};
    return join "\n", @lines;
  }
}

sub _readline {
  my ( $self, @args ) = @_;

  _e('API::Invalid::Params')->throw('_readline() takes no parameters') if @args;

  my $fh = $self->_fh;
  $self->_restore_pos();
  if (wantarray) {
    my @result = <$fh>;
    $self->_set_pos();
    return @result;
  }
  my $result = <$fh>;
  $self->_set_pos();
  return $result;
}

sub _read {
  my ( $self, undef, $len, $offset ) = @_;

  ## no critic ( ProhibitMagicNumbers )
  _e('API::Invalid::Params')->throw('_read() takes 2 or 3 parameters.')
    if ( scalar @_ < 3 or scalar @_ > 4 );

  $self->_restore_pos();
  my $return;
  if ( defined $offset ) {
    $return = read $self->_fh, $_[1], $len, $offset;
  }
  else {
    $return = read $self->_fh, $_[1], $len;
  }
  $self->_set_pos();
  return $return;
}

sub _getc {
  my ($self) = @_;
  _e('API::Invalid::Params')->throw('_get() takes 0 parameters.')
    if scalar @_ > 1;
  $self->_restore_pos();
  my $return = getc $self->_fh;
  $self->_set_pos();
  return $return;
}

sub _seek {
  my ( $self, $position, $whence ) = @_;

  ## no critic ( ProhibitMagicNumbers )

  _e('API::Invalid::Params')->throw('_seek() takes 2 params.')
    if scalar @_ != 3;

  my $fh = $self->_stash->{filehandle};

  if ( 0 == $whence ) {
    $position = $self->_stash->{start_offset} + $position;
  }
  elsif ( 1 == $whence ) {
    $whence   = 0;
    $position = $self->_stash->{current_offset} + $position;
  }
  elsif ( 2 == $whence ) {
  }
  else {
    _e('API::Invalid::Whence')->throw('Expected whence values are 0,1,2');
  }
  my $return = seek $fh, $position, $whence;
  $self->_set_pos();
  return $return;
}

sub _tell {
  my ($self) = shift;
  _e('API::Invalid::Params')->throw('_tell() takes no params.') if @_;
  return $self->_stash->{current_offset} - $self->_stash->{start_offset};
}

sub _eof {
  my $self = shift;
  _e('API::Invalid::Params')->throw("_eof() takes no params : @_ ")
    if @_ && $_[0] != 1;
  $self->_restore_pos();
  return eof $self->_stash->{filehandle};
}

sub _restore_pos {
  my $self = shift;
  return seek $self->_stash->{filehandle}, $self->_stash->{current_offset}, 0;
}

sub _set_pos {
  my $self = shift;
  return ( $self->_stash->{current_offset} = tell $self->_stash->{filehandle} );
}

sub _stash  { return ${ *{ $_[0] } }{stash} }
sub _fileno { return }
sub _e      { return 'Data::Handle::Exception::' . shift }
sub _fh     { return shift->_stash->{filehandle} }

sub _binmode {
  return _e('API::NotImplemented')->throw('_binmode() is difficult on Data::Handle and not implemented yet.');
}

sub _open {
  return _e('API::Invalid')->throw('_open() is invalid on Data::Handle.');
}

sub _close {
  return _e('API::Invalid')->throw('_close() is invalid on Data::Handle');
}

sub _printf {
  return _e('API::Invalid')->throw('_printf() is invalid on Data::Handle.');
}

sub _print {
  return _e('API::Invalid')->throw('_print() is invalid on Data::Handle.');
}

sub _write {
  return _e('API::Invalid')->throw('_write() is invalid on Data::Handle.');
}











1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Handle - A Very simple interface to the __DATA__  file handle.

=head1 VERSION

version 1.000002

=head1 SYNOPSIS

    package Foo;

    sub bar {
        my $handle = Data::Handle->new( __PACKAGE__ );
        while (<$handle>) {
            print $_;
        }
    }

    __DATA__
    Foo

=head1 DESCRIPTION

This Package serves as a very I<very> simple interface to a packages __DATA__ section.

Its primary purposes is to make successive accesses viable without needing to
scan the file manually for the __DATA__ marker.

It does this mostly by recording the current position of the file handle on
the first call to C<< ->new >>, and then re-using that position on every successive C<< ->new >> call,
which eliminates a bit of the logic for you.

At present, it only does a simple heuristic ( backtracking ) to verify the current position is B<immediately>
at the start of a __DATA__ section, but we may improve on this one day.

=head1 METHODS

=head2 new

    my $fh = Data::Handle->new( $targetpackage )

Where C<$targetpackage> is the package you want the __DATA__ section from.

=head1 WARNING

At present, this module does you no favors if something else earlier has moved the file handle position past
the __DATA__ section, or rewound it to the start of the file. This is an understood caveat, but nothing else
seems to have a good way around this either. ( You can always rewind to the start of the file and use heuristics, but that is rather pesky ).

Hopefully, if other people B<do> decide to go moving your file pointer, they'll use this module to do it so
you your code doesn't break.

=head1 USAGE

C<Data::Handle->new()> returns a tied file-handle, and for all intents and purposes, it should
behave as if somebody had copied __DATA__ to its own file, and then done C<< open $fh, '<' , $file >>
on it, for every instance of the Data::Handle.

It also inherits from L<IO::File>, so all the methods it has that make sense to use should probably work
on this too,  i.e.:

    my $handle = Data::Handle->new( __PACKAGE__ );
    my @lines = $handle->getlines();

Also, all offsets are proxied in transit, so you can treat the file-handle as if byte 0 is the first byte of the data section.

    my $handle = Data::Handle->new( __PACKAGE__ );
    my @lines = $handle->getlines();
    seek $handle, 0, 0;
    local $/ = undef;
    my $line = scalar <$handle>; # SLURPED!

Also, the current position of each handle instance is internally tracked, so you can have as many
objects pointing to the same __DATA__ section but have their read mechanism uninterrupted by any others.

    my $handlea  = Data::Handle->new( __PACKAGE__ );
    my $handleb  = Data::Handle->new( __PACKAGE__ );

    seek $handlea, 10, 0;
    seek $handleb, 15, 0;

    read $handlea, my $buf, 5;

    read $handleb, my $bufa, 1;
    read $handleb, my $bufb, 1;

     $bufa eq $bufb;

Don't be fooled, it does this under the covers by a lot of C<seek>/C<tell> magic, but they shouldn't be a problem unless you are truly anal over speed.

=head1 CREDITS

Thanks to LeoNerd and anno, from #perl on irc.freenode.org,
they were most helpful in helping me grok the magic of C<tie> that
makes the simplicity of the interface possible.

Thanks to Chas Owens and James Wright for their efforts with trying to get something simpler with fdup()ing the descriptor ( Sadly not working yet ).

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Data/Handle/IO.pm
use 5.006;
use strict;
use warnings;

package Data::Handle::IO;

our $VERSION = '1.000002';

# ABSTRACT: A Tie Package so Data::Handle can look and feel like a normal handle.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY
















sub TIEHANDLE {
  my $self          = shift;
  my $handle_object = shift;
  return bless $handle_object, $self;
}

sub _object {
  my $self = shift;
  return $self->{self};
}

## no critic (ProtectPrivateSubs)

sub READLINE { return shift->_object->_readline(@_) }
sub READ     { return shift->_object->_read(@_) }
sub GETC     { return shift->_object->_getc(@_) }
sub WRITE    { return shift->_object->_write(@_) }
sub PRINT    { return shift->_object->_print(@_) }
sub PRINTF   { return shift->_object->_printf(@_) }
sub EOF      { return shift->_object->_eof(@_) }
sub CLOSE    { return shift->_object->_close(@_) }
sub BINMODE  { return shift->_object->_binmode(@_) }
sub OPEN     { return shift->_object->_open(@_) }
sub FILENO   { return shift->_object->_fileno(@_) }
sub SEEK     { return shift->_object->_seek(@_) }
sub TELL     { return shift->_object->_tell(@_) }

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Handle::IO - A Tie Package so Data::Handle can look and feel like a normal handle.

=head1 VERSION

version 1.000002

=head1 DESCRIPTION

This is an internal component used by L<Data::Handle> used as a C<tie>
target to provide accessibility to the Perl Core functions, in order to
truly emulate a file-handle.

All the methods on this tie are essentially proxy methods that feed back to
L<Data::Handle> methods, so that all internal calls and all method calls can be coded the same way.

For instance: C<getc($fh)> maps to being the same as if you'd done C< $fh->_getc() >

You're not really supposed to use this package Directly though.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Data/Handle/Exception.pm
use 5.008;    # _use_carp_version
use strict;
use warnings;

package Data::Handle::Exception;

our $VERSION = '1.000002';

# ABSTRACT: Super-light Weight Dependency Free Exception base.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY






























use overload '""' => \&stringify;
use Scalar::Util qw( blessed );
use Carp 1.22;
use Term::ANSIColor qw( YELLOW GREEN RESET );

if ( not defined &Carp::caller_info ) {    ## no critic (Subroutines)
  Carp::croak(<<"EOF");
Cannot load Data::Handle::Exception as your version of Carp does not have
 ::caller_info which we use for backtraces.
Carp Version: $Carp::VERSION
EOF
}








sub new {
  my ($class) = @_;
  my $self = {};
  bless $self, $class;
  return $self;
}







sub throw {
  my $self = shift;

  if ( not blessed $self ) {
    $self = $self->new();
  }
  my $message = shift;

  my @stack      = ();
  my @stacklines = ();

  # This is mostly because want to benefit from all new fixes in carp.
  my $callerinfo = \&Carp::caller_info;    ## no critic (Subroutines)

  {                                        # stolen parts  from Carp::ret_backtrace
    my ($i) = 0;

    my $tid_msg = q{};
    if ( defined &threads::tid ) {         ## no critic (Subroutines)

      my $tid = threads->tid;
      $tid_msg = " thread $tid" if $tid;
    }

    my %i = $callerinfo->($i);

    push @stack, \%i;
    push @stacklines, sprintf q{Exception '%s' thrown at %s line %s%s}, blessed($self), $i{file}, $i{line}, $tid_msg;

    while ( my %j = $callerinfo->( ++$i ) ) {
      push @stack, \%j;
      push @stacklines, sprintf q{%s called at %s line %s%s}, $j{sub_name}, $j{file}, $j{line}, $tid_msg;
    }
  }
  $self->{message}    = $message;
  $self->{stacklines} = \@stacklines;
  $self->{stack}      = \@stack;
  Carp::confess($self);
}

{
  ## no critic ( RequireInterpolationOfMetachars )
  my $s = q{(\x2F|\x5c)};
  my $d = q{\x2E};
  ## use critic
  my $yellow = qr{
      ${s}Try${s}Tiny${d}pm
      |
      ${s}Test${s}Fatal${d}pm
  }x;
  my $green = qr{
    ${s}Data${s}Handle${d}pm
    |
    ${s}Data${s}Handle${s}
  }x;

  sub _color_for_line {
    my $line = shift;
    return YELLOW if ( $line =~ $yellow );
    return GREEN  if ( $line =~ $green );
    return q{};
  }
}


































sub stringify {
  ## no critic ( ProhibitPunctuationVars )
  local $@ = undef;    # Term::ANSIColour clobbers $@
  my $self       = shift;
  my $message    = $self->{message};
  my @stacklines = @{ $self->{stacklines} };

  my $out       = $message . "\n\n";
  my $throwline = shift @stacklines;
  $out .= _color_for_line($throwline) . $throwline . RESET;
  my $i = 2;
  for (@stacklines) {
    $out .= "\n " . _color_for_line($_) . "$i.  " . $_ . RESET;
    $i++;
  }
  return $out . "\n\n";
}
my $dynaexceptions = { 'Data::Handle::Exception' => 1 };

sub _gen {
  my ( undef, $fullclass, $parent ) = @_;
  ## no critic ( RequireInterpolationOfMetachars )
  my $code = sprintf q{package %s; our @ISA=("%s"); 1;}, $fullclass, $parent;

  ## no critic ( ProhibitStringyEval RequireCarping ProhibitPunctuationVars )
  eval $code or throw(qq{ Exception generating exception :[ $@ });
  $dynaexceptions->{$fullclass} = 1;
  return 1;
}

sub _gen_tree {
  my ( $self, $class ) = @_;
  my $parent = $class;

  $parent =~ s{
     ::[^:]+$
    }{}x;
  if ( !exists $dynaexceptions->{$parent} ) {
    $self->_gen_tree($parent);
  }
  if ( !exists $dynaexceptions->{$class} ) {
    $self->_gen( $class, $parent );
  }
  return $class;
}

for (qw( API::Invalid API::Invalid::Whence API::Invalid::Params API::NotImplemented Internal::BadGet NoSymbol BadFilePos )) {
  __PACKAGE__->_gen_tree("Data::Handle::Exception::$_");
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Data::Handle::Exception - Super-light Weight Dependency Free Exception base.

=head1 VERSION

version 1.000002

=head1 SYNOPSIS

    use Data::Handle::Exception;
    Data::Handle::Exception->generate_exception(
        'Foo::Bar' => 'A Bar error occurred :('
    )->throw();

=head1 DESCRIPTION

L<Data::Handle>'s primary goal is to be somewhat "Infrastructural" in design, much like L<Package::Stash> is, being very
low-level, and doing one thing, and doing it well, solving an issue with Perl's native implementation.

The idea is for more complex things to use this, instead of this using more complex things.

As such, a dependency on something like Moose would be overkill, possibly even detrimental to encouraging the use of this module.

So we've scrimped and gone really cheap ( for now at least ) in a few places to skip adding downstream dependencies,
so this module is a slightly nasty but reasonably straight forward exception class.

The actual Exception classes don't actually have their own sources, they're automatically generated when
C<Data::Handle::Exception> is loaded.
And we have some really nice backtraces stolen from Carp's code, with some sexy colored formatting. See L</stringify> for
details.

=head1 METHODS

=head2 new

    my @stack;
    my $i = Data::Handle::Exception->new(  $messageString, \@stack );

=head2 throw

    Data::Handle::Exception->new(  $messageString, \@stack )->throw();

=head2 stringify

Turns this stacktrace into a string.

    $exception->stringify();

    my $str = "hello " . $exception . " world";

If you have a colored terminal, then L<Term::ANSIColor> is used to highlight lines based on how likely they are to be relevant
to diagnosis.

=over 4

=item Green - From Data::Handle and is likely to be "safe", its where the error is being reported from,
so its useful information, but the problem is probably elsewhere.

=item Yellow - Sources we're confident its unlikely to be a source of problems, currently

=over 4

=item Try::Tiny

=item Test::Fatal

=back

=item White - Everything Else, the place the problem is most likely to be.

=back

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentnl@cpan.org>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4210"
  },
  "configure_requires" => {
    "Module::Build" => "0.4210"
  },
  "dist_abstract" => "A Very simple interface to the __DATA__  file handle.",
  "dist_author" => [
    "Kent Fredric <kentnl\@cpan.org>"
  ],
  "dist_name" => "Data-Handle",
  "dist_version" => "0.02001004",
  "license" => "perl",
  "module_name" => "Data::Handle",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => "1.22",
    "IO::File" => 0,
    "Package::Stash" => "0.15",
    "Scalar::Util" => 0,
    "Symbol" => 0,
    "Term::ANSIColor" => 0,
    "Try::Tiny" => 0,
    "overload" => 0,
    "parent" => 0,
    "perl" => "5.008",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Data::Dumper" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001006",
    "lib" => 0
  }
);


my %fallback_build_requires = (
  "Data::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "Module::Build" => "0.4210",
  "Test::Fatal" => 0,
  "Test::More" => "1.001006",
  "lib" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
